ocaml/testsuite/tests/capsule-api/data_atomic.ml

```diff
@@ -0,0 +1,99 @@
+(* TEST
+ include stdlib_alpha;
+ flags += "-extension-universe alpha";
+ runtime5;
+ { bytecode; }
+ { native; }
+*)
+
+(** Testing that capsules can use atomic values *)
+
+module Capsule = Stdlib_alpha.Capsule
+
+type 't myref = { mutable v : 't }
+
+module Atomic : sig
+  type !'a t : atomically_mutable_data
+
+  val make : 'a -> 'a t @@ portable coordinate_nothing
+  val get : 'a t @ coordinated_read -> 'a @@ portable coordinate_nothing
+  val set : 'a t -> 'a -> unit @@ portable coordinate_nothing
+end = struct
+  type !'a t : atomically_mutable_data
+
+  external make : 'a -> 'a t @@ portable coordinate_nothing = "%makemutable"
+  external get : 'a t @ coordinated_read -> 'a @@ portable coordinate_nothing = "%atomic_load"
+  external ignore : 'a -> unit @@ portable coordinate_nothing = "%ignore"
+  external exchange : 'a t -> 'a -> 'a @@ portable coordinate_nothing = "%atomic_exchange"
+  let set r x = ignore (exchange r x)
+end
+
+let a @ portable coordinated_write = Atomic.make 42
+
+type 'a guarded =
+  | Mk : 'k Capsule.Mutex.t * ('a, 'k) Capsule.Data.t -> 'a guarded
+
+let with_guarded x
+  (f : 'k . 'k Capsule.Password.t @ local
+            -> ('a, 'k) Capsule.Data.t
+            -> 'b) =
+  let (Mk (m, p)) = x in
+  Capsule.Mutex.with_lock m (fun k -> f k p)
+;;
+
+let map_with_guarded x
+  (f : 'k . 'k Capsule.Password.t @ local
+            -> ('a, 'k) Capsule.Data.t
+            -> ('b, 'k) Capsule.Data.t) =
+  let (Mk (m, p)) = x in
+  Mk (m, Capsule.Mutex.with_lock m (fun k -> f k p))
+;;
+
+let atomic_then_id @ portable coordinate_writing =
+  fun i w ->
+    Atomic.set a i; w
+
+(* [Data.create] *)
```

Comment:
Some comments above each test mentioning what property is being tested would be useful